File: src/lib/prn.js
// ---------------------------------------------------------------------------
// Números aleatorios permanentes (PRN) — Ohlsson 1995.
//
// Cada concepto recibe un número fijo en [0, 1) derivado de SHA-256(semilla ‖ '|' ‖ id).
// Está en la muestra si su número es menor que la fracción dentro de su estrato (dominio).
// Como el número no depende de qué otros conceptos existan, la muestra de hoy es
// subconjunto de la de mañana: los conceptos nuevos entran o no por su propio número y
// los ya valorados no se mueven. Ampliar un dominio es subir la fracción.
//
// pipeline/prn.py implementa EXACTAMENTE lo mismo; tests/prn.test.js comprueba que los
// dos dan los mismos números sobre vectores fijos.
// ---------------------------------------------------------------------------

const HEX = 13 // 13 hex = 52 bits, lo que cabe exacto en un double

export async function prn(semilla, id) {
  const datos = new TextEncoder().encode(`${semilla}|${id}`)
  const hash = await crypto.subtle.digest('SHA-256', datos)
  const hex = [...new Uint8Array(hash)].map((b) => b.toString(16).padStart(2, '0')).join('')
  return parseInt(hex.slice(0, HEX), 16) / Math.pow(16, HEX)
}

// Selección estratificada proporcional con suelo. `conceptos`: [{id, dominio, prn}].
// Devuelve el conjunto de ids incluidos por el estrato aleatorio.
export function muestrear(conceptos, fraccion, suelo) {
  const porDominio = new Map()
  for (const c of conceptos) {
    if (!porDominio.has(c.dominio)) porDominio.set(c.dominio, [])
    porDominio.get(c.dominio).push(c)
  }
  const incluidos = new Set()
  for (const [, lista] of porDominio) {
    const ordenada = [...lista].sort((a, b) => a.prn - b.prn || (a.id < b.id ? -1 : 1))
    const porFraccion = ordenada.filter((c) => c.prn < fraccion)
    const elegidos = porFraccion.length >= suelo ? porFraccion : ordenada.slice(0, Math.min(suelo, ordenada.length))
    for (const c of elegidos) incluidos.add(c.id)
  }
  return incluidos
}

File: pipeline/prn.py
"""Números aleatorios permanentes (PRN) — la misma función que src/lib/prn.js.

    prn(semilla, id) -> float en [0, 1)

SHA-256 de "semilla|id", los 13 primeros hex (52 bits) divididos por 16**13. Trece y no
más porque es lo que un double de JavaScript representa exacto: así el número que calcula
el pipeline y el que pueda calcular el navegador son idénticos bit a bit.
"""
from __future__ import annotations

import hashlib

HEX = 13


def prn(semilla: str, id_: str) -> float:
    h = hashlib.sha256(f"{semilla}|{id_}".encode("utf-8")).hexdigest()
    return int(h[:HEX], 16) / 16 ** HEX


def muestrear(conceptos: list[dict], fraccion: float, suelo: int) -> set[str]:
    """Estratificado proporcional por dominio, con suelo. conceptos: [{id, dominio, prn}]."""
    por_dominio: dict[str, list[dict]] = {}
    for c in conceptos:
        por_dominio.setdefault(c["dominio"], []).append(c)
    incluidos: set[str] = set()
    for lista in por_dominio.values():
        ordenada = sorted(lista, key=lambda c: (c["prn"], c["id"]))
        por_fraccion = [c for c in ordenada if c["prn"] < fraccion]
        elegidos = por_fraccion if len(por_fraccion) >= suelo else ordenada[: min(suelo, len(ordenada))]
        incluidos.update(c["id"] for c in elegidos)
    return incluidos


if __name__ == "__main__":  # vectores de prueba para tests/prn.test.js
    import json
    import sys

    semilla = sys.argv[1] if len(sys.argv) > 1 else "prueba"
    ids = sys.argv[2:] or ["CPT-00001", "CPT-00002", "CPT-04030"]
    print(json.dumps({i: prn(semilla, i) for i in ids}, indent=2))
